File: Python/algorithmic_heights/24_med.py
# Median
import random
import sys


def convert_to_intlist(line):
    return [int(item) for item in line.split()]


def three_way_partitioning(array, pivot, start_ix, end_ix):
    current_ix = start_ix
    while current_ix <= end_ix:
        elem = array[current_ix]
        if elem < pivot:
            if current_ix != start_ix:
                array[current_ix], array[start_ix] = array[start_ix], elem
            current_ix += 1
            start_ix += 1
        elif elem > pivot:
            array[current_ix], array[end_ix] = array[end_ix], elem
            end_ix -= 1
        else:
            current_ix += 1
    return start_ix, end_ix + 1


def find_kth_smallest_element(array, n, k):
    start_ix = 0
    end_ix = n - 1
    kth_smallest = None
    while kth_smallest is None:
        random_ix = random.randint(start_ix, end_ix)
        pivot = array[random_ix]
        middle_start, middle_end = three_way_partitioning(array, pivot, start_ix, end_ix)
        if k <= middle_start:
            end_ix = middle_start
        elif middle_start < k <= middle_end:
            kth_smallest = array[middle_start]
        else:
            start_ix = middle_end
    return kth_smallest


def main():
    reader = sys.stdin
    n = int(next(reader))
    array = convert_to_intlist(next(reader))
    k = int(next(reader))
    result = find_kth_smallest_element(array, n, k)
    print(result)


if __name__ == '__main__':
    main()

File: Haskell/algorithmic_heights/24_med.hs
import Control.Monad.ST (runST)
import Data.Vector (Vector, fromList, thaw)
import System.Random (randomR, mkStdGen)
import qualified Data.Vector as V
import qualified Data.Vector.Mutable as MV

convertToIntList :: String -> [Int]
convertToIntList = map read . words


findKthSmallestElement :: Vector Int -> Int -> Int -> Int
findKthSmallestElement vector n k =
    let threeWayPartitioning array pivot startIx endIx =
            let swapElems currentIx startIx endIx =
                    if currentIx > endIx then return (startIx, endIx + 1)
                    else do
                        elem <- MV.read array currentIx
                        if elem < pivot then do
                            MV.swap array currentIx startIx
                            swapElems (currentIx + 1) (startIx + 1) endIx
                        else if elem > pivot then do
                            MV.swap array currentIx endIx
                            swapElems currentIx startIx (endIx - 1)
                        else swapElems (currentIx + 1) startIx endIx
            in swapElems startIx startIx endIx

        go array startIx endIx gen = do
            let (randomIx, gen') = randomR (startIx, endIx) gen
            pivot <- MV.read array randomIx
            (middleStart, middleEnd) <- threeWayPartitioning array pivot startIx endIx
            if k <= middleStart then go array startIx middleStart gen'
            else if k <= middleEnd then MV.read array middleStart
            else go array middleEnd endIx gen'

    in runST $ do
        array <- thaw vector
        let gen = mkStdGen 2112
        go array 0 (n - 1) gen


main :: IO ()
main = do
    n <- readLn
    xs <- fromList . convertToIntList <$> getLine
    k <- readLn
    print $ findKthSmallestElement xs n k
